src/ocean/net/ssl/SslClientConnection.d

```diff
@@ -0,0 +1,622 @@
+        // support AF_UNSPEC.
+
+        auto addr_err = this.addr_info.get(host_name, host_port, AF_UNSPEC,
+            SOCK_STREAM, 0);
+
+        if ( addr_err != 0 )
+        {
+            this.ssl_error(host_name, "connect", gai_strerror(addr_err));
+        }
+
+        this.socket.socket(addr_info.info().ai_family,
+            addr_info.info().ai_socktype | SocketFlags.SOCK_NONBLOCK,
+            addr_info.info().ai_protocol);
+
+        if ( !socket.connect(addr_info.info().ai_addr) )
+        {
+            this.error(host_name, "socket.connect", "Failed to connect");
+        }
+
+        this.sslHandle = SSL_new(globalSslContext);
+
+        if ( !this.sslHandle )
+        {
+            this.error(host_name, "connect", "sslNew failed");
+        }
+
+        if (!SSL_set_fd(this.sslHandle, this.socket.fileHandle))
+        {
+            this.error(host_name, "connect", "sslSetFd failed");
+        }
+
+        // Set it into "connect" state, ie it is a client, not a server.
+        // (For a server, SSL_set_accept_state would be called instead).
+
+        SSL_set_connect_state(this.sslHandle);
+
+        ERR_clear_error();
+
+        int r;
+
+        while ( (r = SSL_do_handshake(this.sslHandle)) != 1 )
+        {
+            int err = SSL_get_error(this.sslHandle, r);
+
+            if ( err ==  SSL_ERROR_WANT_WRITE )
+            {
+                this.select_client.ioWait(Event.EPOLLOUT);
+            }
+            else if (err == SSL_ERROR_WANT_READ)
+            {
+                this.select_client.ioWait(Event.EPOLLIN);
+            }
+            else
+            {
+                this.error(host_name, "connect", "Handshake failed");
+            }
+        }
+    }
+
+
+    /***************************************************************************
+
+        Write a string to the SSL connection
+
+        The calling task is blocked until the write has completed.
+
+        Params:
+            request = the string to send
+
+        Throws:
+            IOException if the write failed
+
+    ***************************************************************************/
+
+    public void write ( cstring request )
+    {
+        int w;
+
+        while ( (w = SSL_write(this.sslHandle, cast(void *)request.ptr,
+            cast(int)request.length)) <= 0 )
+        {
+            auto err = SSL_get_error(this.sslHandle, w);
+
+            // At any time, a renegotiation is possible, so a call to SSL_write
+            // can also cause read operations.
+
+            if ( err ==  SSL_ERROR_WANT_WRITE )
+            {
+                this.select_client.ioWait(Event.EPOLLOUT);
+            }
+            else if (err == SSL_ERROR_WANT_READ)
+            {
+                this.select_client.ioWait(Event.EPOLLIN);
+            }
+            else
+            {
+                // The write failed
+
+                this.ssl_error("", "write", ERR_reason_error_string(err));
+            }
+        }
```

Comment:
It looks that this code for handing the epoll integration is same in all cases. Can it be moved to a single method?

Reply:
I'm not sure if the code is fundamentally the same, or if it just looks the same by accident.
If the socket is blocking, the return values of `SSL_write` are different to `SSL_read`, and `SSL_do_handshake` is completely different, and even when non-blocking, the error codes other than `SSL_ERROR_WANT_READ` and `SSL_ERROR_WANT_WRITE`, are different for reads, writes, and handshakes. It might be necessary to treat them differently in the future.

It was also a non-obvious feature of the code. Many examples which I saw, didn't realize that you need to do this for writes, as well as for reads. The fact that you might need to do a write while you are doing a read, is highly counter-intuitive. Whereas it's totally expected for handshaking.

Since this pattern only appears in 3 places (and no other SSL calls behave this way), I don't think it's possible to reduce code size by attempting to extract it.

Reply:
> the error codes other than SSL_ERROR_WANT_READ and SSL_ERROR_WANT_WRITE, are different for reads, writes, and handshakes

OK, this is a good point, and I agree good point for keeping it as is.

Reply:
(it was not visible from the code I've seen).

Reply:
What you could do is to add
```d
private void doSslIo ( lazy int io, void delegate(int err) handle_error )
{
    while (true)
    {
        int w = io;
        if (w > 0)
            break;

        auto err = SSL_get_error(this.sslHandle, w);
        // At any time, a renegotiation is possible, so a call to SSL_write
        // can also cause read operations.
        switch (err)
        {
            case SSL_ERROR_WANT_WRITE:
                this.select_client.ioWait(Event.EPOLLOUT);
                break;
            case SSL_ERROR_WANT_READ:
                this.select_client.ioWait(Event.EPOLLIN);
                break;
            default:
                handle_error(err);
        }
    }
}
```
and call it in `connect`, `read` and `write`.

Reply:
The interesting thing that Nemanja has noticed, is that the `ioWait` calls are actually the only place that relies on epoll, or tasks. So, if those `select_client.ioWait` calls were abstracted, it would be possible to use this code in the existing swarm implementation, for example.
This gives two competing refactorings for the same piece of code.